src/main/sensors/compass.c

```diff
@@ -388,31 +414,241 @@ uint32_t compassUpdate(timeUs_t currentTimeUs)
     }
 
     flightDynamicsTrims_t *magZero = &compassConfigMutable()->magZero;
-    if (magInit) {              // we apply offset only once mag calibration is done
-        mag.magADC[X] -= magZero->raw[X];
-        mag.magADC[Y] -= magZero->raw[Y];
-        mag.magADC[Z] -= magZero->raw[Z];
-    }
 
     if (tCal != 0) {
-        if ((currentTimeUs - tCal) < 30000000) {    // 30s: you have 30s to turn the multi in all directions
+        if (cmpTimeUs(tCal, currentTimeUs) > 0) {
             LED0_TOGGLE;
-            for (int axis = 0; axis < 3; axis++) {
-                if (mag.magADC[axis] < magZeroTempMin.raw[axis])
-                    magZeroTempMin.raw[axis] = mag.magADC[axis];
-                if (mag.magADC[axis] > magZeroTempMax.raw[axis])
-                    magZeroTempMax.raw[axis] = mag.magADC[axis];
+            // collect relevant data for calibration routine
+            const float frequency = 1.0f / (dTimeUs * 1.0e-6f);
+            const float dmag[3] =  {(mag.magADC[X] - magPrevious.magADC[X]) * frequency,
+                                    (mag.magADC[Y] - magPrevious.magADC[Y]) * frequency,
+                                    (mag.magADC[Z] - magPrevious.magADC[Z]) * frequency};
+            const float gyroADCfRadians[3] =  {DEGREES_TO_RADIANS(gyro.gyroADCf[X]),
+                                               DEGREES_TO_RADIANS(gyro.gyroADCf[Y]),
+                                               DEGREES_TO_RADIANS(gyro.gyroADCf[Z])};
+            // it is assumed that the user has started to move the multi if squared norm of rotational speed vector is greater than GYRO_NORM_SQUARED_MIN
+            const float gyroNormSquared = sq(gyroADCfRadians[X]) + sq(gyroADCfRadians[Y]) + sq(gyroADCfRadians[Z]);
+            if (!didMovementStart && gyroNormSquared > GYRO_NORM_SQUARED_MIN) {
+                didMovementStart = true;
+                // starting now, the user has CALIBRATION_TIME_US to move the multi in a figure of eight in all directions
+                tCal = micros() + CALIBRATION_TIME_US;
+            }
+            // only start calibration after the user has started to move the multi
+            if(didMovementStart) {
+                compassBiasEstimatorApply(&compassBiasEstimator, mag.magADC, dmag, gyroADCfRadians);
             }
         } else {
             tCal = 0;
             for (int axis = 0; axis < 3; axis++) {
-                magZero->raw[axis] = (magZeroTempMin.raw[axis] + magZeroTempMax.raw[axis]) / 2; // Calculate offsets
+                magZero->raw[axis] = lrintf(compassBiasEstimator.b[axis]);
             }
-
             saveConfigAndNotify();
         }
     }
 
+    magPrevious = mag;
+
+    if (magInit) {
```

Comment:
(not related; it seems that magInit is newer false)

Reply:
Yeah, will remove it.